ctapipe/image/time_next_neighbor_cleaning.py

```diff
@@ -0,0 +1,742 @@
+import numpy as np
+from scipy.spatial import cKDTree as KDTree
+from scipy.interpolate import interp1d
+
+import astropy.units as u
+import warnings
+import json
+
+
+def get_n_order_neighbor_matrix(geometry, order):
+    """
+    Get a matrix of the first and second order neighbors for each pixel.
+
+    Parameters
+    ----------
+    geometry
+
+    Returns
+    -------
+    Boolean matrix of dimension (n_pix, n_pix) with the first and second
+    neighbors for all pixels.
+    """
+    neighbors = geometry.neighbor_matrix_sparse
+    for i in range(order - 1):
+        neighbors = neighbors.dot(geometry.neighbor_matrix_sparse)
+    neighbors = neighbors.toarray()
+    np.fill_diagonal(neighbors, 0)
+    return neighbors
+
+
+def dilate_matrix(mask, matrix):
+    return np.dot(matrix, mask)
+
+
+def check_number_of_neighbors(mask, geo, min_neighbors=3, order=2, keep_neighbors=True):
+    neighbor_matrix = get_n_order_neighbor_matrix(geo, order)
+    n_neighbors = np.sum(mask * neighbor_matrix, axis=1) - 1
+
+    valid_pixels = n_neighbors >= min_neighbors
+    if keep_neighbors:
+        valid_pixels = dilate_matrix(valid_pixels, neighbor_matrix)
+
+    return valid_pixels & mask
+
+
+class TimeNextNeighborCleaning:
```

Comment:
Should this be a subclass of ImageCleaner? I guess so

Reply:
Or at least there should be a wrapper that is a subclass of ImageCleaner that calls this class internally. That would allow you to not only specify options from a config file, but also lets the user choose the image cleaner on the fly. See how other ImageCleaners work.

Reply:
Change to `TimeNextNeighborCleaning(ImageCleaner)`